python/ray/autoscaler/v2/scheduler.py

```diff
@@ -417,15 +516,223 @@ def schedule(self, request: SchedulingRequest) -> SchedulingReply:
+
+        # Update the context
+        terminating_nodes.extend(to_terminate_nodes)
+        assert len(all_nodes) == len(
+            terminating_nodes + non_terminating_nodes
+        ), "The number of nodes should be the same after enforcing max nodes."
+
+        all_nodes = terminating_nodes + non_terminating_nodes
+        self._ctx.update(all_nodes)
+        logger.debug(
+            "After enforced max nodes for global num nodes limit : {}".format(self._ctx)
+        )
+
+    @staticmethod
+    def _terminate_nodes(
+        nodes: List[SchedulingNode],
+        num_to_terminate: int,
+        cause: TerminationRequest.Cause,
+        max_num_nodes: Optional[int] = None,
+        max_num_nodes_per_type: Optional[int] = None,
+        idle_duration_ms: Optional[int] = None,
+    ) -> Tuple[List[SchedulingNode], List[SchedulingNode]]:
+        """
+        Terminate the nodes.
+
+        Args:
+            nodes: The nodes to be terminated.
+            num_to_terminate: The number of nodes to be terminated.
+            cause: The cause of the termination.
+
+            max_num_nodes: The max number of nodes for the entire cluster only
+                used when the cause is TerminationRequest.Cause.MAX_NUM_NODES.
+            max_num_nodes_per_type: The max number of nodes for each node type.
+                Only used when the cause is
+                TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE.
+            idle_duration_ms: The idle duration in ms. Only used when the cause is
+                TerminationRequest.Cause.IDLE.
+
+        Returns:
+            A tuple of:
+                - The terminated nodes.
+                - The remained nodes.
+        """
+
+        terminated_nodes, remained_nodes = (
+            nodes[:num_to_terminate],
+            nodes[num_to_terminate:],
+        )
+
+        for node in terminated_nodes:
+            node.status = SchedulingNodeStatus.TO_TERMINATE
+            node.termination_request = TerminationRequest(
+                id=str(time.time_ns()),
+                instance_id=node.im_instance_id,
+                ray_node_id=node.ray_node_id,
+                cause=cause,
+            )
+            if cause == TerminationRequest.Cause.MAX_NUM_NODES:
+                node.termination_request.max_num_nodes = max_num_nodes
+            elif cause == TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE:
+                node.termination_request.max_num_nodes_per_type = max_num_nodes_per_type
+            elif cause == TerminationRequest.Cause.IDLE:
+                node.termination_request.idle_duration_ms = idle_duration_ms
+            else:
+                raise ValueError("Unknown termination cause: {}".format(cause))
+
+        return terminated_nodes, remained_nodes
+
+    @staticmethod
+    def _sort_nodes_for_termination(node: SchedulingNode) -> Tuple:
+        """
+        Sort the nodes for termination increasingly by:
+
+            1. First if it doesn't have a ray node.
```

Comment:
first the node is not allocated or ray hasn't been started..?  not sure what exactly `First if it doesn't have a ray node.` mean